### ISSUE TITLE
@@unique block attribute fails for non-scalar types (regression)

### ISSUE THREAD

**Issue by ssalbdivad** [prisma/prisma2#1440]
Based on the following snippet from [the PSL spec](https://github.com/prisma/specs/tree/master/schema#why-not-support-id-for-multiple-blocks), ```@@unique``` should work fine for non-scalar types:

```@@unique([recipe, ingredient])``` would mean that the combination of fields must be unique in the table:
```
model RecipeIngredient {
  recipe                            Recipe
  ingredient                      Ingredient
  amount                          Float
  quantitativeDisclosures String
  comment                       String
  @@unique([recipe, ingredient])
}
```

The semantic meaning seems intuitive here and implies a comparison based on the ```@id``` field of the ```Recipe``` and ```Ingredient``` objects. As of ```2.0.0-preview017.2```, this was working but some time after that point started failing with a panic as described in #1352. #1352 was closed in favor of #1377, but @janpio mentioned #1377 covers improving the error message but not the regression.

I know nothing about Rust, but https://github.com/prisma/prisma-engine/pull/241 looks like a likely root cause to me. The timing is about right and it seems like all of the tests expect that fields passed to ```@@unique``` are scalar.

**Comment by janpio** [comment]
> but @janpio mentioned #1377 covers improving the error message but not the regression.

You misunderstood me there, #1377 is tracking the fixing of that problem.

Your issue unfortunately does not include any information what exactly is going wrong, and how stuff is failing, but I assume you are either introspecting your database or trying to use an existing schema and then getting an error message similar to the one noted in #1377?

Your initial comment additionaly asked a different question:

> Is `@@unique` only supposed to be used for scalar fields now?

No, it is not as far as I am aware. I would assume that currently the functionality is not fully implemented, which causes error messages like the one seen in #1377.

@pantharshit00 will be able to confirm what I wrote above I hope. 
He might also be able to give more insight into the "regression" part of what you mentioned, that it worked in a previous version.

**Comment by ssalbdivad** [comment]
I'm having a really hard time following what you're asking for. In #1352 (logged on 12/18), @PaiJi wrote a detailed description of the problem including a full stack trace. I'll copy it in my next comment for reference.

I commented on that issue the next day on 12/19:

> @pantharshit00 This blocks me from upgrading from 2.0.0-preview017.2 as well and still exists as of 2.0.0-preview019. It seems to be related to the `@@unique` operator.

On 1/20, @divyenduz closed the issue and commented:
> Closing this in favor of #1377

In general, when an issue is closed "in favor" of another, I assume if I want to follow up, I should comment on the new issue. So I did, but you then told me:

> Hey @ssalbdivad, can you please open a new issue for this? I get the connection to this issue, but the original is really just a bug tracking issue for this specific problem with this specific error message. Thanks!

Since I had already indicated I was having the same issue @PaiJi described, I figured you must have meant #1377 had to do with the error message itself as opposed to the root cause of the error. If the goal of #1377 is to resolve the exact error @PaiJi described, then we should close this issue and move its description back to a comment on #1377 because it's exactly the same issue I was trying to investigate. My comment indicated a possibility that there had been a change in the spec since #1377's description includes:

> Panic: Model (probably) has a @@id or @@unique, that refers to a non-scalar field. This usually happens in table inheritance and join tables without IDs.

Which implies ```@@unique``` doesn't allow non-scalar fields anymore.

Finally, you mentioned:

> I assume you are either introspecting your database or trying to use an existing schema

I encountered the issue with a fresh database created by prisma. It's easily reproducible-- just include ```@@unique(nonScalarField, someOtherField)``` in a model.

**Comment by ssalbdivad** [comment]
Stack trace posted by @PaiJi in #1352 (also applies to this issue):

> Hello developers, i am using prisma2 to build my project , recently after upgrade prisma2 **version >2.0.0-preview017.1** i can't building my project anymore and keep get this error info:
> 
> ```
> Error: Schema parsing
> thread 'main' panicked at 'Unable to resolve scalar field 'owner'', src/libcore/option.rs:1190:5
> stack backtrace:
>    0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
>    1: core::fmt::write
>    2: std::io::Write::write_fmt
>    3: std::panicking::default_hook::{{closure}}
>    4: std::panicking::default_hook
>    5: std::panicking::rust_panic_with_hook
>    6: std::panicking::continue_panic_fmt
>    7: rust_begin_unwind
>    8: core::panicking::panic_fmt
>    9: core::option::expect_failed
>   10: prisma_models::index::IndexTemplate::build::{{closure}}
>   11: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
>   12: prisma_models::index::IndexTemplate::build
>   13: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
>   14: prisma_models::model::ModelTemplate::build
>   15: <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
>   16: prisma_models::internal_data_model::InternalDataModelTemplate::build
>   17: prisma::cli::CliCommand::execute
>   18: <std::future::GenFuture<T> as core::future::future::Future>::poll
>   19: tokio::runtime::global::with_thread_pool
>   20: tokio::runtime::global::with_state
>   21: tokio::runtime::spawner::Spawner::enter
>   22: tokio::runtime::time::variant::with_default
>   23: std::thread::local::LocalKey<T>::with
>   24: prisma::main
>   25: std::rt::lang_start::{{closure}}
>   26: std::panicking::try::do_call
>   27: __rust_maybe_catch_panic
>   28: std::rt::lang_start_internal
>   29: main
> ```
> 
> field 'owner' in model is foreign key, and used in unique like this
> 
> ```
> model ExchangeRate {
>   id           String   @default(cuid()) @id
>   fromCurrency String
>   toCurrency   String
>   owner        User
> 
>   @@unique([fromCurrency, toCurrency, owner])
> }
> ```
> 
> didn't something i wrote wrong?

**Comment by janpio** [comment]
> My comment indicated a possibility that there had been a change in the spec since #1377's description includes:
> 
> > Panic: Model (probably) has a @@id or @@unique, that refers to a non-scalar field. This usually happens in table inheritance and join tables without IDs.
> 
> Which implies `@@unique` doesn't allow non-scalar fields anymore.

Yes, that is obviously the case right now because you get an error. But that is not the intended behavior, which is why there is a bug that is being tracked in #1377. We obviously want to get rid of that panic, and the limitation that causes to our users.

The sentence you quote is an internal description of the behavior during Introspection, which helps us to understand what is going on.

Does this help you understand what is going on?

**Comment by ssalbdivad** [comment]
@janpio My only goal with regard to any of this is the resolution of the error originally mentioned in #1352. If that is also a goal of #1377, then I don't understand the purpose of this being its own issue. 

> It does not imply wanted behavior or a change to any specification

```Panic: Model (probably) has a @@id or @@unique, that refers to a non-scalar field. This usually happens in table inheritance and join tables without IDs.``` does not make any sense as an error message, internal or not, unless its implication is that ```@@unique``` and ```@@id``` are not allowed to refer to non-scalar fields.

The goal of my original comment on #1377 was to aid in the investigation process by suggesting that there may be a misalignment between the spec as it is documented and the one the developer actually built. The fact that there are several tests for ```@@unique``` in prisma/prisma-engine#241, none of which reference a non-scalar field, supports that theory as well. Even if none of that is true, I'm not describing a distinct issue-- I'm just wrong :upside_down_face:

**Comment by janpio** [comment]
> > It does not imply wanted behavior or a change to any specification
> 
> `Panic: Model (probably) has a @@id or @@unique, that refers to a non-scalar field. This usually happens in table inheritance and join tables without IDs.` does not make any sense as an error message, internal or not, unless its implication is that `@@unique` and `@@id` are not allowed to refer to non-scalar fields.

I have to disagree here, it does. (I created it, so I obviously have to think so... 🙈)

Let me give you the full background: 
As indicated in the issue this is an pattern description of the Introspection Analysis. This is a process where we run `prisma introspect` on a number of database schemas, and see if that proceeds or not. We collect all the error messages, and then sort them into "patterns" (basically: types or errors) which have a description. This sentence is one of them. They are meant for us internally to better understand why this Introspection is failing.

Some schemas obviously have UNIQUE indexes for multiple columns, some of which might actually be foreign keys. When these are represented in Prisma Schemas via Introspection, this results in a model that looks exactly like yours: `@@unique([field1, field2])`. As the Introspection also includes a validation step, we also get the exact same error you are getting (`thread 'main' panicked at 'Unable to resolve scalar field 'field1''`). As our goal in the Introspection Analysis is to get rid of all of these (obviously), similar errors are collected (basically all the `Unable to resolve scalar field` ones, no matter what the field name is) and the pattern `panic-unable-to-resolve-scalar-field` is applied. That means they are grouped in our internal system and attributed to the same root cause.

Did this convince you that this description can make sense?

This pattern and the underlying error are then tracked in an Github issue which can be used to deduplicate and close all issues that basically tell us about the same problem - so every one that has the `Unable to resolve scalar field` error message because of an `@@unique`, like #1352, #1410 or #1405.

And then the next step is to actually fix the bug, make the panic go away and close the issue and make everyone happy. We are not there yet for this specific bug though.

> The goal of my original comment on #1377 was to aid in the investigation process by suggesting that there may be a misalignment between the spec as it is documented and the one the developer actually built. The fact that there are several tests for @@unique in prisma/prisma-engine#241, none of which reference a non-scalar field, supports that theory as well. Even if none of that is true, I'm not describing a distinct issue-- I'm just wrong 🙃

Ok, here I can offer 2 responses:
1. No spec mismatch, just a bug that needs to be fixed.
2. prisma-engine#241 is just an internal code change, leading up to bigger things. There will need to be other changes before #1377 can be closed which will then obviously include code and tests to support @@unique for scalar fields.

**Comment by janpio** [comment]
Ah, what I am a bit surprised (but believe you - but was not aware of) is that this used to work up until `2.0.0-preview017.2` and actually really enforced this. It is moot though to investigate why and how, as fixing #1377 really will reenable this functionality.

By the way, I appreciate you spending time on your Sunday afternoon having a discussion with my about this 🚀

**Comment by ssalbdivad** [comment]
@janpio Thanks for the detailed explanation. Let me know if I have this right:

```Panic: Model (probably) has a @@id or @@unique, that refers to a non-scalar field. This usually happens in table inheritance and join tables without IDs.``` describes a set of conditions that include no inherent problems per the spec, but the message was written after it was already a known bug that ```@@unique``` breaks with non-scalar input. Its goal is to let devs know "Hey, you're probably seeing this because your model has a non-scalar ```@@unique``` or ```@@id```, but it's not your fault and it's a known issue."

That would make sense to me, although maybe linking to the corresponding GitHub issue as part of the message would be useful for anyone who sees this and other similar messages, e.g.:

```Panic: Model (probably) has a @@id or @@unique... Tracked by https://github.com/prisma/prisma2/issues/1377```

I can imagine as either a dev on the team looking to quickly understand/communicate about a known issue they run into or an external contributor trying to understand an error message that doesn't seem like it should be an error at all (like me :blush:), that could be a big help.

I do think it's worth considering the possibility that there may be confusion among the team about whether ```@@unique``` is expected to handle non-scalar fields. As I mentioned, I've taken a look at several PRs that include tests for ```@@unique```, including prisma/prisma-engine#241, and none of them test non-scalar fields. In fact, the only reference I've found in the spec to non-scalar fields being passed to ```@@unique``` is a [subsection of the FAQ titled "Why not support @id for multiple blocks?"](https://github.com/prisma/specs/tree/master/schema#why-not-support-id-for-multiple-blocks), which doesn't seem centrally concerned with clarifying this. Maybe everyone on the team, including @dpetrick who worked on prisma/prisma-engine#241, is totally aligned around the expected behavior here and non-scalar fields are just failing due to a dev mistake, but in that case it seems important to add at least one dedicated example to the PSL spec and add tests verifying the behavior.

Clearly at some point someone who worked on this thought non-scalar fields should be allowed, as I doubt them working just happened by accident :smile: I do think it's important to note that this is a regression and should probably be prioritized so that users who have been waiting to upgrade to 20 can do so. I also still think (admittedly based on little domain knowledge) that prisma/prisma-engine#241 is a likely candidate as the timing is near when the regression occurred and (again, not knowing Rust) it seems like a major refactor of @@unique to support prisma client queries, which I can imagine affecting how scalar/non-scalar fields are handled, and again it didn't have any tests for non-scalar fields. If that's all totally wrong and doesn't help with solving #1377, then I apologize for wasting your time, but I would love to see this fixed so I can upgrade to preview 20 so am doing my best to help without learning another programming language :stuck_out_tongue: 

Happy to try to work through this on a Sunday; I've been spending the rest of the morning working on a ```nexus-prisma``` feature anyways, so the context switching's not too bad :bowtie:

**Comment by janpio** [comment]
> Let me know if I have this right:
> 
> `Panic: Model (probably) has a @@id or @@unique, that refers to a non-scalar field. This usually happens in table inheritance and join tables without IDs.` describes a set of conditions that include no inherent problems per the spec, but the message was written after it was already a known bug that `@@unique` breaks with non-scalar input. Its goal is to let devs know "Hey, you're probably seeing this because your model has a non-scalar `@@unique` or `@@id`, but it's not your fault and it's a known issue."

Yes, one could describe it like that.

> although maybe linking to the corresponding GitHub issue as part of the message would be useful for anyone who sees this and other similar messages, e.g.:
> 
> `Panic: Model (probably) has a @@id or @@unique... Tracked by https://github.com/prisma/prisma2/issues/1377`
> 
> I can imagine as either a dev on the team looking to quickly understand/communicate about a known issue they run into or an external contributor trying to understand an error message that doesn't seem like it should be an error at all (like me 😊), that could be a big help.

Got that covered, the description usually appears together with a link to #1377. That part of course did not make much sense to mention in #1377 itself ;)

> I do think it's worth considering the possibility that there may be confusion among the team about whether @@unique is expected to handle non-scalar fields. As I mentioned, I've taken a look at several PRs that include tests for @@unique, including prisma/prisma-engine#241, and none of them test non-scalar fields. In fact, the only reference I've found in the spec to non-scalar fields being passed to @@unique is a subsection of the FAQ titled "Why not support @id for multiple blocks?" , which doesn't seem centrally concerned with clarifying this. Maybe everyone on the team, including @dpetrick who worked on prisma/prisma-engine#241 , is totally aligned around the expected behavior here and non-scalar fields are just failing due to a dev mistake, but in that case it seems important to add at least one dedicated example to the PSL spec and add tests verifying the behavior.

I can assure you there is no confusion in the team. `@@unique` should just work the way you expect it to, and tbh there is no reason why scalar and non scalar fields should behave differently - they do not in MySQL/Postgres/SQLite.

You are not seeing anything about `@@unique` and scalar fields because that work is not yet finished and committed. But people are working on things related to this, and as soon as the code is there you should see appropriate tests and code changes.

The unclarity about specs are indeed true though, I would appreciate if you could create an issue over at https://github.com/prisma/specs so this is explicitly tracked. (It is on my own list to go through that document, but having specific pointers for deficits would make this easier.)

> I do think it's important to note that this is a regression and should probably be prioritized so that users who have been waiting to upgrade to 20 can do so. I also still think (admittedly based on little domain knowledge) that prisma/prisma-engine#241
is a likely candidate as the timing is near when the regression occurred and (not knowing Rust) it seems like a major refactor of @@unique to support prisma client queries, which I can imagine affecting how scalar/non-scalar fields are handled, and again it didn't have any tests for non-scalar fields. 

#1377 will get fixed, as mentioned before the team is actively working on fixing some of the underlying problems right now, so it is a kind of cosmetic point if it is a bug or a regression.

But if you want, you can of course create a minimal reproduction (1 db table, so 1 model in schema, simple script that shows error message on second execution) with `2.0.0-preview017.2` that shows this feature working (and put it on GitHub and post the link here). We can use and updated version (quite some changes in naming from preview017.2 to today...) of that to double check that we fixed the problem when #1377 is finally closed.

> If that's all totally wrong and doesn't help with solving #1377, then I apologize for wasting your time, but I would love to see this fixed so I can upgrade to preview 20 so am doing my best to help without learning another programming language 😛

I think in this case you would actually also have to be hired by us before being able to really change anything (vs. "just" learning Rust), as the problem why this does not work right now goes _very_ deep into our codebase - which is also the reason why we could not just fix this in the next version after we broke it.

---

Our discussion by the way made me rethink how we structure this problem in GitHub issues right now (it was presented as an Introspection problem, but is actually a Query Engine problem - which only _also_ manifests in Introspection), and I might spend some time moving some stuff around tomorrow - if my colleagues agree ;) Then I would probably classify the Query Engine issue as a regression (hence your reproduction being useful).

**Comment by janpio** [comment]
I looked into the regression part a bit, and I think I might have an explanation how this came to be:

1. Up to `2.0.0-preview017.2` `@@unique` was actually not a concept known to the Query Engine (which runs your queries from the client), as it was just enforced from the databases itself (`UNIQUE` constraint on two columns, errors on failure, done). 
1. Then https://github.com/prisma/prisma-engine/pull/241 added a way to use the `@@unique` definition to write specific queries which we wanted to have. For very complicated internal reasons the Query Engine could only do that for non scalar fields, and bamm we have your problem - code that worked before does not any more. 
1. Shortly after work on fixing this began, but is a huge thing that touches not only `@@unique` but some of the fundamental engine concepts that were in place so this work is still ongoing (and will also affect many of the other remaining bugs like the `@@id` stuff tracked in https://github.com/prisma/prisma2/issues/1083)
1. I will be a very happy person when these changes are merged and both #1377 and #1083 can be closed.

**Comment by ssalbdivad** [comment]
@janpio Created the spec issue at https://github.com/prisma/specs/issues/387.

I'm glad you were able to understand what I meant by calling the issue a "regression" and trace it back to prisma/prisma-engine#241. It definitely helps me feel less insane based on all the evidence I thought I found that things changed when that PR was released :smile:

I get how from the team's perspective it might seem like that is a new feature in query-engine and not a regression, but I'm sure you understand how to a (dedicated) external user, based on the facts that:

1. ```@@unique``` was already referenced in the docs
2. ```@@unique``` on a non-scalar went from giving a reasonable error like "Violated uniqueness constraint" on a bad query to blocking the entire prisma workflow.

This really seems like a regression. I understand these kinds of trade-offs are sometimes required and that prisma2 is still in beta, but maybe some end-to-end tests for stuff like this would help catch unexpected side-effects to developer experience. In this case, it might have even been feasible to build in some temporary logic so that query-engine continued to ignore ```@@unique``` for non-scalar fields while supporting it for scalars ```@@unique``` since I'm guessing it's pretty easy to differentiate between scalar and non-scalar fields when parsing PSL.

**Comment by janpio** [comment]
The bug/regression question is mostly a cosmetic thing, and based on the issue handling we are following right now labelling #1377 a regression would not be correct (the Introspection problem is new, after all). But as I wrote before this discussion showed me that this might not be enough, and if I make some changes tomorrow, we could have #1377 for Introspection and a new issue for the `@@unique` defect which will be labeled as a `kind/regression`.

Unfortunately this will not influence when the fix for that will be released ;) (No ETA on that yet - code is being written for it every day though)

> In this case, it might have even been feasible to build in some temporary logic so that query-engine continued to ignore `@@unique` for non-scalar fields while supporting it for scalars `@@unique` since I'm guessing it's pretty easy to differentiate between scalar and non-scalar fields when parsing PSL.

We tend to not do that, as this could apply to _many_ changes (very often some previous behavior is changed or removed that _someone_ might rely on) and would lead to a collection of upcoming changes that would then all be released at the same time (when the magical last change is done that unblocks all of the ones that were held back). Whatever goes wrong then is super hard to understand and differentiate between the individual changes.

But I get where you are coming from and why you would think this would be a good idea - it would solve your problem after all. In the future I might have this conversation in my mind when having to inform a decision how to deal with a similar problem.

Have you actually looked into (and are interested in) workarounds? I almost think you could just leave the `@@unique` out of your `schema.prisma` file and still define it on the database level, and everything might work as expected. (Might cause problems if you use `prisma2 migrate --experimental` to migrate your database of course)

**Comment by ssalbdivad** [comment]
@janpio I've been using ```migrate``` for a while to maintain a SQLite DB for local dev and a Postgres DB in AWS, and I've been extremely thankful for the extent to which prisma has abstracted away my need to write SQL for either :pray:

> `We tend to not [...][want to spend time on feature toggles], as this could apply to many changes (very often some previous behavior is changed or removed that someone might rely on) and would lead to a collection of upcoming changes that would then all be released at the same time (when the magical last change is done that unblocks all of the ones that were held back). Whatever goes wrong then is super hard to understand and differentiate between the individual changes.`

It sounds like somewhere in the prisma ecosystem, there is logic responsible for parsing a ```@@unique``` constraint and telling the DB how to create a table that implements it. I was a happily consuming that logic until the ```query-engine``` change decided it should care about ```@@unique```, even with non-scalar fields it doesn't support. I am totally obsessed with writing maintainable code, understand your concerns, and would share your assessment of the situation if it required an anomalous temporary hack. However, building a simple filter to allow parts of a prisma schema to be recognized differently by ```query-engine``` and ```migrate``` would not be throwaway code. On the contrary, for every feature that currently exists that is supported by ```migrate``` but not by the ```query-engine```, as well as every PSL feature that requires ```migrate``` and ```query-engine``` code in the future (and potentially other PSL parsing entities I don't know about), this kind of feature toggle could provide a trivial way to get functionality that is ready in the hands of users sooner (it seems like in this case the ```migrate``` functionality was implemented a few months ago and the ```query-engine``` is still being worked on) and provide an escape hatch for changes like this that would otherwise break a user's workflow. This pattern is extremely common and many large codebases find a way to implement it extremely maintainably; it's only when there's no cohesive strategy that it becomes untenable. Would recommend [this article about feature toggles](https://martinfowler.com/articles/feature-toggles.html) if you're interested.

**Comment by janpio** [comment]
Week has gone, I did not get to update the issues yet. 
Work is still ongoing on the problem itself but we are making progress.

**Comment by janpio** [comment]
This is now cleaned up with additional issue https://github.com/prisma/prisma2/issues/1534 which is marked as a regression.
